<a href="https://colab.research.google.com/github/rubymanderna/ML_ECGR5105/blob/main/Homework_3_ML/Homework_3_4_and_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Problem 4 (20pts):

Use the cancer dataset to build a logistic regression model to classify the type of cancer (Malignant vs. benign). Use the PCA feature extraction for your training. Perform N number of independent training (N=1, …, K). Identify the optimum number of K, principal components that achieve the highest classification accuracy. Report your classification accuracy, precision, recall, and F1 score over a different number of Ks. Explain and elaborate on your results and compare them against problems 2 and 3.

In [8]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [9]:
# Load the dataset
df = pd.read_csv("/content/drive/MyDrive/ML_ECGR5105/cancer.csv")
print(df.columns)
# Step 1: Data Preprocessing
df = df.drop(columns=['id', 'Unnamed: 32'])
df['diagnosis'] = df['diagnosis'].map({'M': 1, 'B': 0})

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32'],
      dtype='object')


In [10]:


X = df.drop(columns=['diagnosis'])
y = df['diagnosis']


# Normalize the features (min-max scaling)
scaler = MinMaxScaler()
X = scaler.fit_transform(X)

# Initialize variables
best_accuracy = 0
best_k = 0

# Loop for different numbers of Principal Components (K)
for k in range(1, 11):  # Adjust range as needed
    # Step 3: PCA
    pca = PCA(n_components=k)
    X_pca = pca.fit_transform(X)

    # Step 4: Model Training
    X_train, X_test, y_train, y_test = train_test_split(X_pca, y, test_size=0.2, random_state=42)
    model = LogisticRegression()
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    # Step 5: Performance Metrics
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)

    # Store metrics
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_k = k

    # Report metrics for each K
    print(f'K = {k}: Accuracy = {accuracy}, Precision = {precision}, Recall = {recall}, F1 Score = {f1}')

# Step 6: Identify Optimum K
print(f'Best K: {best_k}, Best Accuracy: {best_accuracy}')


K = 1: Accuracy = 0.9473684210526315, Precision = 0.9743589743589743, Recall = 0.8837209302325582, F1 Score = 0.9268292682926831
K = 2: Accuracy = 0.9649122807017544, Precision = 1.0, Recall = 0.9069767441860465, F1 Score = 0.951219512195122
K = 3: Accuracy = 0.9736842105263158, Precision = 0.9761904761904762, Recall = 0.9534883720930233, F1 Score = 0.9647058823529412
K = 4: Accuracy = 0.9649122807017544, Precision = 0.975609756097561, Recall = 0.9302325581395349, F1 Score = 0.9523809523809524
K = 5: Accuracy = 0.9824561403508771, Precision = 1.0, Recall = 0.9534883720930233, F1 Score = 0.9761904761904763
K = 6: Accuracy = 0.9824561403508771, Precision = 1.0, Recall = 0.9534883720930233, F1 Score = 0.9761904761904763
K = 7: Accuracy = 0.9824561403508771, Precision = 1.0, Recall = 0.9534883720930233, F1 Score = 0.9761904761904763
K = 8: Accuracy = 0.9824561403508771, Precision = 1.0, Recall = 0.9534883720930233, F1 Score = 0.9761904761904763
K = 9: Accuracy = 0.9824561403508771, Precisi

Problem 5 (20pts):

Can you repeat problem 4? This time, replace the logistic regression classifier with the Bayes classifier. Report your results (classification accuracy, precision, recall and F1 score). Compare your results against problems 2, 3 and 4.

In [13]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score


# Initialize variables
best_accuracy = 0
best_k = 0

# Assuming X is your feature matrix
scaler = MinMaxScaler()  # Or StandardScaler
X_normalized = scaler.fit_transform(X)

# Loop for different numbers of Principal Components (K)
for k in range(1, 11):  # Adjust range as needed
    # Step 3: PCA

    pca = PCA(n_components=k)
    X_pca = pca.fit_transform(X_normalized)

    # Step 4: Naive Bayes Classifier (Gaussian)
    X_train, X_test, y_train, y_test = train_test_split(X_pca, y, test_size=0.2, random_state=42)
    model = GaussianNB()
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    # Step 5: Performance Metrics
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)

     # Store metrics
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_k = k

    # Report metrics for each K
    print(f'K = {k}: Accuracy = {accuracy}, Precision = {precision}, Recall = {recall}, F1 Score = {f1}')

# Step 6: Identify Optimum K
print(f'Best K: {best_k}, Best Accuracy: {best_accuracy}')

K = 1: Accuracy = 0.956140350877193, Precision = 0.975, Recall = 0.9069767441860465, F1 Score = 0.9397590361445783
K = 2: Accuracy = 0.956140350877193, Precision = 0.975, Recall = 0.9069767441860465, F1 Score = 0.9397590361445783
K = 3: Accuracy = 0.956140350877193, Precision = 0.975, Recall = 0.9069767441860465, F1 Score = 0.9397590361445783
K = 4: Accuracy = 0.956140350877193, Precision = 0.975, Recall = 0.9069767441860465, F1 Score = 0.9397590361445783
K = 5: Accuracy = 0.956140350877193, Precision = 0.975, Recall = 0.9069767441860465, F1 Score = 0.9397590361445783
K = 6: Accuracy = 0.9473684210526315, Precision = 0.9512195121951219, Recall = 0.9069767441860465, F1 Score = 0.9285714285714286
K = 7: Accuracy = 0.9473684210526315, Precision = 0.9302325581395349, Recall = 0.9302325581395349, F1 Score = 0.9302325581395349
K = 8: Accuracy = 0.9473684210526315, Precision = 0.9302325581395349, Recall = 0.9302325581395349, F1 Score = 0.9302325581395349
K = 9: Accuracy = 0.9473684210526315, 

**Explaination for 4 and 5**



**Final result from problem 2**

**After adding weight penalty**

Final Accuracy: 0.9736842105263158
Final Precision: 1.0
Final Recall: 0.9302325581395349
Final F1 Score: 0.963855421686747
Confusion Matrix: [[71 0]
                                 [ 3 40]]
The addition of a penalty term helped to regularize our model and prevent overfitting up to some extent (very slightly but yes) confusion matrix- in Normalize method we have 0 false positive cases, However, we have False negative increase by 1 but on the other hand true negative decreased by 1, our desired goal for cancer data should be to decrease the number of False negative results, Based on confusion matrix better model was without adding penalties(this is based on cancer scenario assumptions).


**final result from problem 3**

Accuracy: 0.9649122807017544

Precision: 0.975609756097561

Recall: 0.9302325581395349

F1 Score: 0.9523809523809524

compare - in our case logistic regression classifier is prformin better for all the metrics.However,if model is overfitting then using naive bias would be good choice here.

**final result from problem 4**

Best K : 5

Best Accuracy : 0.9824561403508771

PCA is a technique used for dimensionality reduction.
The first few principal components capture the most variance in the data thus reducing dimensionality of the dataset
for this case 5 components will capture the most important information in data.

These 5 principal components are linear combinations of the original features. They are selected in such a way that they explain the maximum variance in the dataset.

In terms of accuracy PCA performs better and also it helps in reducing dimensionality of the dataset

**final result from problem 5**

Best K : 1

Best Accuracy : 0.956140350877193

Clearly the model is not performing good if we refer problem 4,3,2 results in terms of accuracy

Also it is suggesting that most of the variance in data is coming from one feature which is kind off not possible.
